usr/src/uts/armv8/os/fakebop.c

```diff
@@ -332,6 +415,328 @@ do_bsys_alloc(bootops_t *bop, caddr_t virthint, size_t size, int align)
+
+	/*
+	 * use a bootprop_t at curr_page and link into list
+	 */
+	b = (bootprop_t *)curr_page;
+	curr_page += sizeof (bootprop_t);
+	curr_space -=  sizeof (bootprop_t);
+	b->bp_next = bprops;
+	bprops = b;
+
+	/*
+	 * follow by name and ending zero byte
+	 */
+	b->bp_name = curr_page;
+	bcopy(name, curr_page, nlen);
+	curr_page += nlen;
+	*curr_page++ = 0;
+	curr_space -= nlen + 1;
+
+	/*
+	 * set the property type
+	 */
+	b->bp_flags = flags & DDI_PROP_TYPE_MASK;
+
+	/*
+	 * copy in value, but no ending zero byte
+	 */
+	b->bp_value = curr_page;
+	b->bp_vlen = vlen;
+	if (vlen > 0) {
+		bcopy(value, curr_page, vlen);
+		curr_page += vlen;
+		curr_space -= vlen;
+	}
+
+	/*
+	 * align new values of curr_page, curr_space
+	 */
+	while (curr_space & 0xf) {
+		++curr_page;
+		--curr_space;
+	}
+}
+
+static void
+bsetprops(char *name, char *value)
+{
+	bsetprop(DDI_PROP_TYPE_STRING, name, strlen(name),
+	    value, strlen(value) + 1);
+}
+
+static void
+bsetprop32(char *name, uint32_t value)
+{
+	bsetprop(DDI_PROP_TYPE_INT, name, strlen(name),
+	    (void *)&value, sizeof (value));
+}
+
+static void
+bsetprop64(char *name, uint64_t value)
+{
+	bsetprop(DDI_PROP_TYPE_INT64, name, strlen(name),
+	    (void *)&value, sizeof (value));
+}
+
+static void
+bsetpropsi(char *name, int value)
+{
+	char prop_val[32];
+
+	(void) snprintf(prop_val, sizeof (prop_val), "%d", value);
+	bsetprops(name, prop_val);
+}
+
+/*
+ * to find the size of the buffer to allocate
+ */
+int
+do_bsys_getproptype(bootops_t *bop __unused, const char *name)
+{
+	bootprop_t *b;
+
+	for (b = bprops; b != NULL; b = b->bp_next) {
+		if (strcmp(name, b->bp_name) != 0)
+			continue;
+		return (b->bp_flags);
+	}
+
+	return (-1);
+}
+
+/*
+ * to find the size of the buffer to allocate
+ */
+int
+do_bsys_getproplen(bootops_t *bop __unused, const char *name)
+{
+	bootprop_t *b;
+
+	for (b = bprops; b; b = b->bp_next) {
+		if (strcmp(name, b->bp_name) != 0)
+			continue;
+		return (b->bp_vlen);
+	}
+
+	return (-1);
+}
+
+/*
+ * get the value associated with this name
+ */
+int
+do_bsys_getprop(bootops_t *bop __unused, const char *name, void *value)
+{
+	bootprop_t *b;
+
+	for (b = bprops; b; b = b->bp_next) {
+		if (strcmp(name, b->bp_name) != 0)
+			continue;
+		bcopy(b->bp_value, value, b->bp_vlen);
+		return (0);
+	}
+
+	return (-1);
+}
+
+/*
+ * get the name of the next property in succession from the standalone
+ */
+static char *
+do_bsys_nextprop(bootops_t *bop __unused, char *name)
+{
+	bootprop_t *b;
+
+	/*
+	 * A null name is a special signal for the first boot property
+	 */
+	if (name == NULL || strlen(name) == 0) {
+		if (bprops == NULL)
+			return (NULL);
+		return (bprops->bp_name);
+	}
+
+	for (b = bprops; b; b = b->bp_next) {
+		if (name != b->bp_name)
+			continue;
+
+		b = b->bp_next;
+		if (b == NULL)
+			return (NULL);
+
+		return (b->bp_name);
+	}
+
+	return (NULL);
+}
+
+/*
+ * Parse numeric value from a string. Understands decimal, hex, octal, - and ~
+ */
+static int
```

Comment:
Could we use `ddi_strtoul` and then just handle the radix and `~` ?

Reply:
Yeah, that seems reasonable. Again, what's here is just a transposition of what's in i86pc.

I do think there's a ton that can be shared between the platforms (from this sort of code all the way to most of the PCI code).